cmd/bblfshd/main.go

```diff
@@ -94,7 +94,12 @@ func main() {
 	logrus.Infof("bblfshd version: %s (build: %s)", version, build)
 
 	r := buildRuntime()
-	d := daemon.NewDaemon(version, r, buildGRPCOptions()...)
+	grpcOpts, err := cmdutil.GRPCSizeOptions(*maxMessageSize)
+	if err != nil {
+		logrus.Warnln(err)
```

Comment:
I also think this is the right place to fail and exit the program.

Reply:
@juanjux @dennwc but what is the advantage of doing that?

In current ([documented](https://github.com/bblfsh/sdk/pull/329/files#diff-3d26f06e2a3da8bfb583a3f488a3baafR28)) behavior if this happens, we just cap the limit to 100MiB, report it to the user as warning and continue running.

I belive this has the advantage of much better user experience - it does the right thing recovering automatically from this simple error, falling back to reasonable defaults and letting user know about that.

I'm open to arguments of benefits of previous approach though!

Reply:
To address arguments from https://github.com/bblfsh/sdk/pull/329#discussion_r235762041

@dennwc 
> This is an error on the program startup

yes, but it is a recoverable one - so we recover. I'm trying to understand what is the benefit of moving cognitive load of doing that to the user, if software can do it with a reasonable default? So far, I can not see

@juanjux @creachadair 
> it's using grpcOpts obtained from a call that returned an error, it definitely should exit the program or at least the function.

Why?  For me "Idiomatic code" <<<<< "User Experience" and plus, as current behavior is not idiomatic, it's well documented.

Reply:
We should not force ourselves to choose between user experience and idiomatic code. The reason we have this difficult now, as far as I can tell, is that we have used error-handing to propagate what should have been a diagnostic log.

My view is we should not try to resolve this problem here and now, but take a more systematic approach to separating places where we've used errors to indicate "unfortunate" rather than "erroneous" conditions. Then we can go back and clean these things up. A proper solution will probably (as @bzz pointed out) require some changes to logger propagation.

Reply:
How often do we expect to hit this limit in practice? Is it often enough that we should consider adding chunking (on the request side) and pagination (on the response)? 

In other words: If it's really important to the caller to set a specific value here (e.g., because they have huge files and can't get work done), should we make the API support that without requiring the user to think about gRPC internals? 

This again is a larger question than we can solve right here in this PR, but I'm trying to get an understanding of the decision space. Can we avoid forcing the user to think about this at all?

Reply:
Hardcoding the limit and adding a chunking (streaming) endpoint sounds like the best option to me. As a separate PR, of course :)

Reply:
@creachadair 
> We should not force ourselves to choose between user experience and idiomatic code. 

totally agree.

> used error-handing to propagate what should have been a diagnostic log

thank you for ☝️, yes, that is the case now.

@dennwc 
> I don't think it's a recoverable error.
> Let's say the default .... The user tries to sends ..

Thank you for illustrative example @dennwc, it really aids understand and makes 👍argument.

>  As a note, the issue is a bit worse, since the log message will pop up in the log of the driver, not in bblfshd

A bit confused how logs from [`cmd/bblfshd/main.go`](https://github.com/bblfsh/bblfshd/pull/216/files#diff-94bf84897cc9a7746dc1bac15c4484eaR97) can pop up in a driver. Driver gRPC server process has it's own logs, where similar condition was also taken care of. 

Anyway, thanks everyone for productive discussion - I think https://github.com/bblfsh/bblfshd/pull/216#discussion_r235792138 is an overriding argument here, I will change it back to error condition.

On
>> How often do we expect to hit this limit in practice? Is it often enough that we should consider adding chunking (on the request side) and pagination (on the response)?
>
> Hardcoding the limit and adding a chunking (streaming) endpoint sounds like the best option

I think this is going to be quite often, as soon as gitbase is used to process a substantial fraction of Github.

The fact that nobody of the downstream clients in DR has stumbled up on this limitation most probably means that they work it around by only parsing a lines smaller then the limit (a [quick search](https://github.com/src-d/gitbase/blob/master/internal/function/uast.go#L28) verifies this guess)

Having streaming gRPC endpoint sounds like possible solution, but before implementing it I would prefer to have a separate issue \w a proposal and then verify @creachadair concerns with more clients/users first.

WDYT?

Reply:
I think for now we should set the limits high enough to push the tail below some limit of interest. That has the advantage of being something we can do now to unblock other work. By the time you hit 2GiB on the wire, it's perfectly reasonable to report to the user "you're doing something weird" and let them come talk to us.

I agree that a streaming solution is a bigger project and deserves a separate design process.

Reply:
> I think for now we should set the limits high enough

I belive that is exactly what we have done here - it is set 100MiB for now.